text/0000-delegation.md

```diff
@@ -0,0 +1,700 @@
+This unresolved question is relevant to the decision:
+-   Are there any cases of [Custom Self Types](https://github.com/rust-lang/rfcs/pull/2362) where `self` needs to be manually dereferenced, e.g.
+    ```rust
+    impl TR for S {
+        fn foo(self: Arc<Self>) -> u32 {
+            *self.f.foo()
+        }
+    }
+    ```
+
+
+### Omitting the `impl` block
+
+The `impl` block is a lot of extra noise when simply delegating all trait items to a struct field. There can be a lot of generic type parameters which are copy-pasted from the member field's trait impl.
+
+```rust
+delegate TR::* to S::f;
+delegate TR::* to S::0;
+delegate TR::{fn foo, fn bar}
+    to S::f;
+delegate TR::* to S::getter();
+delegate fn push to AppendOnlyVec::0;
+delegate S::* to trait TR;
+```
+
+This is shorter and easier to read than the equivalent:
+
+```rust
+impl TR for S { delegate * to f; }
+impl TR for S { delegate * to 0; }
+impl TR for S {
+    delegate fn foo, fn bar
+        to f;
+}
+impl TR for S { delegate * to getter(); }
+impl<T> AppendOnlyVec<T> { delegate fn push to 0; }
+impl S { delegate * to trait TR; }
+```
+
+However, in the case of delegating most of a trait's methods, an `impl` block is still required and now the "overridden" method can be seperated from the delegate statement, which could be confusing and easier to miss what is happening.
+
+We could offer both syntaxes as a good 4 step ratchet:
+
+1.  first try `#[derive(..)]`
+2.  then go with `delegate TR::* to S::f;`
+3.  then delegate inside an impl
+4.  finally implement things manually.
+
+
+### Assume items are `fn`
+
+When listing items to delegate, the prefix `fn` is assumed, since it is most common. Items may be prefixed with `fn` if desired.
+
+```rust
+impl TR for S {
+    delegate foo, bar, const MAX, type Item
+        to f;
+}
+impl TR for S {
+    delegate to f for foo, bar, const MAX, type Item;
+}
+delegate TR::{foo, bar, const MAX, type Item}
+    to S::f;
+delegate push to AppendOnlyVec::0;
+```
+
+
+### Different syntax for delegating most trait items
+
+```rust
+impl TR for S {
+    delegate _ to f;
+
+    fn foo(&self) -> u32 {
+        42
+    }
+}
+```
+
+This overcomes the drawback: the `delegate *` syntax may be too implicit.
+
+It also overcomes the objection to omitting the `impl` block: the "overridden" method can be seperated from the delegate statement, which could be confusing and easier to miss what is happening.
+
+The increase in cognitive load is minimal, since `_` is widely used to mean "inferred by the compiler."
+
+
+### Other syntax options the authors chose not to use in this RFC:
+
+Many of these syntaxes were never “rejected” in the original RFC’s comment thread and are likely still on the table. This list merely describes the authors' rationale for preferring `delegate ... to field_name;` over all of these alternatives.
+
+-   `impl TR for S use self.F { ... }` was criticized in the first RFC’s comment thread for looking too much like inheritance.
```

Comment:
Why is that a problem? It basically is inheritance. Or a restricted version of it, because inheritance delegates all methods to the designated member called “base” while this only delegates methods of specific trait. But that restriction is obvious from this syntax.

Reply:
"Inheritance" traditionally means a LOT more than simply delegating implementations to another type, such as committing to the same interface, having the same internal structure as another type, or being implicitly convertable to other types. In Rust those things are usually kept separate, which we think is a good thing.

So if we wanted to make this part of the RFC clearer, we could say that `impl TR for S use self.F {}` makes it surprising that a field named `F` must already exist in `S`'s declaration, rather than being implicitly added by this `impl TR for S use self.F {}`. Or we could say that `impl TR for S use self.F {}` makes it surprising that `S` cannot be implicitly converted to an `F`.

I don't personally think there's any one or two specific features in the "inheritance" grabbag that most users would mistakently expect when they see `impl TR for S use self.F {}`, but I do think a lot of people would mistakenly expect _something_ on top of method delegation if they saw that syntax.

Reply:
Inheritance does not mean all that much more than delegation, really. It does mean that all interfaces will be delegated, but it should be obvious it's *not* happening here, because it starts by stating which trait it is delegating. It does *not* mean committing to the same structure, only to containing that structure, but that needs to happen here as well. The only other thing inheritance does is the implicit coercion of reference. That won't happen here, but we only said we are delegating specific trait after all. And coercion to that trait does work.

This form would not be appropriate for delegating inherent methods, because that *would* look like it might be doing more than it does. But for the traits, it would be convenient shortcut that does not really look like promising more than it does.

> makes it surprising that a field named `F` must already exist in `S`'s declaration

Since it does not mention the type of `F`, it's quite clear that that still has to be given.

> makes it surprising that `S` cannot be implicitly converted to an `F`

I would somewhat agree for inherent methods, i.e. `impl S use self.F {}`. But I wouldn't fear that much for the `impl TR for S` case.

Reply:
Hm, but there is one difference that I have to admit would be quite confusing—in case of inheritance, the overridden behaviour applies even if you've got a reference to the “base field”. But in our case it won't. In fact, I fear delegating some methods, but not all of them, will be confusing with any syntax.